cmd/wallet/deposits.go

```diff
@@ -0,0 +1,189 @@
+package walletcmd
+
+import (
+	"encoding/binary"
+	"encoding/hex"
+	"fmt"
+	"os"
+	"sort"
+	"text/tabwriter"
+
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/ethereum/go-ethereum/crypto"
+
+	"github.com/keep-network/keep-core/pkg/bitcoin"
+	"github.com/keep-network/keep-core/pkg/tbtc"
+)
+
+type depositEntry struct {
+	wallet [20]byte
+
+	depositKey      string
+	revealedAtBlock uint64
+	isSwept         bool
+
+	fundingTransactionHash        bitcoin.Hash
+	fundingTransactionOutputIndex uint32
+	amountBtc                     float64
+}
+
+// ListDeposits gets deposits from the chain.
+func ListDeposits(
+	tbtcChain tbtc.Chain,
+	walletStr string,
+	hideSwept bool,
+	sortByAmount bool,
+) error {
+	deposits, err := getDeposits(tbtcChain, walletStr)
+	if err != nil {
+		return fmt.Errorf(
+			"failed to get deposits: [%w]",
+			err,
+		)
+	}
+
+	if len(deposits) == 0 {
+		return fmt.Errorf("no deposits found")
+	}
+
+	// Filter
+	if hideSwept {
+		deposits = removeSwept(deposits)
+	}
+
+	// Order
+	sort.SliceStable(deposits, func(i, j int) bool {
+		return deposits[i].revealedAtBlock > deposits[j].revealedAtBlock
+	})
+
+	if sortByAmount {
+		sort.SliceStable(deposits, func(i, j int) bool {
+			return deposits[i].amountBtc < deposits[j].amountBtc
+		})
+	}
+
+	// Print
+	if err := printTable(deposits); err != nil {
+		return fmt.Errorf("failed to print deposits table: %v", err)
+	}
+
+	return nil
+}
+
+func removeSwept(deposits []depositEntry) []depositEntry {
+	result := []depositEntry{}
+	for _, deposit := range deposits {
+		if deposit.isSwept {
+			continue
+		}
+		result = append(result, deposit)
+	}
+	return result
+}
+
+func getDeposits(tbtcChain tbtc.Chain, walletStr string) ([]depositEntry, error) {
+	logger.Infof("reading deposits from chain...")
+
+	filter := &tbtc.DepositRevealedEventFilter{}
+	if len(walletStr) > 0 {
+		walletPublicKeyHash, err := hexToWalletPublicKeyHash(walletStr)
+		if err != nil {
+			return []depositEntry{}, fmt.Errorf("failed to extract wallet public key hash: %v", err)
+		}
+
+		filter.WalletPublicKeyHash = [][20]byte{walletPublicKeyHash}
+	}
+
+	depositRevealedEvent, err := tbtcChain.PastDepositRevealedEvents(filter)
+	if err != nil {
+		return []depositEntry{}, fmt.Errorf(
+			"failed to get past deposit revealed events: [%w]",
+			err,
+		)
+	}
+
+	logger.Infof("found %d DepositRevealed events", len(depositRevealedEvent))
+
+	result := make([]depositEntry, len(depositRevealedEvent))
+	for i, event := range depositRevealedEvent {
+		logger.Debugf("getting details of deposit %d/%d", i+1, len(depositRevealedEvent))
+
+		depositKey := buildDepositKey(event.FundingTxHash, event.FundingOutputIndex)
+
+		depositRequest, err := tbtcChain.GetDepositRequest(event.FundingTxHash, event.FundingOutputIndex)
+		if err != nil {
+			return result, fmt.Errorf(
+				"failed to get deposit request: [%w]",
+				err,
+			)
+		}
+
+		result[i] = depositEntry{
```

Comment:
> If there is at least one deposit in the proposal that has no required confirmations, the entire proposal will be rejected by wallet members.

This is why we add proposal validation before submitting it in https://github.com/keep-network/keep-core/pull/3549

Reply:
This makes sense and validating before submission makes sense as well. However, I'd like to know from the very beginning if I should even look at the given deposit and consider it for sweeping. This will be also important once we automate this job.

Reply:
Yes, sure, I understood that. It will be added in a follow-up.